repository: added support for tx calls

gushakov opened the pull request:
Modified `FedoraRepositoryImpl` slightly to add some useful calls to RESTfull API to deal with transactions. Based on the description given at https://wiki.duraspace.org/display/FEDORA4x/RESTful+HTTP+API+-+Transactions

The idea is to be able to do something like this.

```java
repo.startTransaction();
try {
  repo.createObject(path);
  // create datastreams, related objects, update properties, etc.
}
catch (Exception e) {
  // roll back if something goes wrong
  repo.rollbackTransaction();
}
repo.commitTransaction();
// or
```

The transaction IDs (with `tx:` prefix) are stored in `ThreadLocal<String>` attribute of the static utility class.

awoods commented:
@gushakov, Thank you for this. Can you please submit an iCLA:
https://wiki.duraspace.org/display/DSP/Contributor+License+Agreements

@mikedurbin or @escowles, I am on travel this week. Will you have some time to review this, and create a [JIRA](https://jira.duraspace.org/browse/FCREPO/) ticket?

escowles commented:
I've created https://jira.duraspace.org/browse/FCREPO-1762 and should have time to review this today or tomorrow.

escowles commented on fcrepo-client-impl/src/main/java/org/fcrepo/client/impl/FedoraRepositoryImpl.java:
```diff
@@ -315,8 +325,110 @@ public boolean isWritable() {
+        // check if path needs to be prepended with a slash
+        final String txId = TransactionIdHolder.getCurrentTransactionId();
+        return (txId == null || path.contains(TX))
+                ? path : (path.startsWith("/") ? "/" + txId + path : txId + "/" + path);
```
I think it would be good to avoid nested ?: conditions -- they are hard enough to read without being nested.

escowles commented on fcrepo-client-impl/src/main/java/org/fcrepo/client/utils/TransactionIdHolder.java:
```diff
@@ -0,0 +1,53 @@
+ */
+package org.fcrepo.client.utils;
+
+/**
```
Does this class provide functionality beyond what using a ThreadLocal<String> variable in FedoraRepositoryImpl.java?

Also, please include a class description and @since statement in the Javadoc comment.

escowles commented on fcrepo-client-impl/src/main/java/org/fcrepo/client/impl/FedoraRepositoryImpl.java:
```diff
@@ -224,7 +233,8 @@ public FedoraObject createObject(final String path) throws FedoraException {
     public FedoraObject createResource(final String containerPath) throws FedoraException {
-        final HttpPost post = httpHelper.createPostMethod(containerPath == null ? "" : containerPath, null);
+        final HttpPost post = httpHelper
+                .createPostMethod(containerPath == null ? "" : prependTransactionId(containerPath), null);
```
Shouldn't we still use the tx even if containerPath is null?

escowles commented on fcrepo-client-impl/src/main/java/org/fcrepo/client/impl/FedoraRepositoryImpl.java:
```diff
@@ -76,8 +83,8 @@ public FedoraRepositoryImpl(final String repositoryURL) {
      * @param repositoryURL Repository base URL
-     * @param username Repository username
-     * @param password Repository password
+     * @param username      Repository username
```
Please rollback whitespace-only changes.

escowles commented on fcrepo-client-impl/src/main/java/org/fcrepo/client/impl/FedoraRepositoryImpl.java:
```diff
@@ -88,7 +95,7 @@ public FedoraRepositoryImpl(final String repositoryURL, final String username, f
      *
      * @param repositoryURL Repository baseURL
-     * @param httpClient Pre-configured httpClient
+     * @param httpClient    Pre-configured httpClient
```
Please rollback whitespace-only changes.

escowles commented on fcrepo-client-impl/src/main/java/org/fcrepo/client/impl/FedoraRepositoryImpl.java:
```diff
@@ -113,7 +120,7 @@ public boolean exists(final String path) throws FedoraException, ForbiddenExcept
                 LOGGER.error("error checking resource {}: {} {}", uri, statusCode, status.getReasonPhrase());
                 throw new FedoraException("error checking resource " + uri + ": " + statusCode + " " +
-                                          status.getReasonPhrase());
+                        status.getReasonPhrase());
```
Please rollback whitespace-only changes.

escowles commented on fcrepo-client-impl/src/main/java/org/fcrepo/client/impl/FedoraRepositoryImpl.java:
```diff
@@ -153,7 +162,7 @@ public FedoraDatastream createDatastream(final String path, final FedoraContent
                 LOGGER.error("error creating resource {}: {} {}", uri, statusCode, status.getReasonPhrase());
                 throw new FedoraException("error retrieving resource " + uri + ": " + statusCode + " " +
-                                                  status.getReasonPhrase());
+                        status.getReasonPhrase());
```
Please rollback whitespace-only changes.

escowles commented on fcrepo-client-impl/src/main/java/org/fcrepo/client/impl/FedoraRepositoryImpl.java:
```diff
@@ -212,7 +221,7 @@ public FedoraObject createObject(final String path) throws FedoraException {
                 LOGGER.error("error creating resource {}: {} {}", uri, statusCode, status.getReasonPhrase());
                 throw new FedoraException("error retrieving resource " + uri + ": " + statusCode + " " +
-                                                  status.getReasonPhrase());
+                        status.getReasonPhrase());
```
Please rollback whitespace-only changes.

escowles commented:
@gushakov: This looks good -- I have a few minor comments, and this looks ready to merge once those are addressed and your CLA is signed.

gushakov commented:
As suggested: 
 - moved `ThreadLocal` tx id holder variable into `FedoraRepositoryImpl` 
 - prepending for the null `containerPath` in `createResource` 
 - rollback whitespace-only changes

Also, removed my name from "author" in javadoc (too small of a contribution).

escowles commented:
:+1:  -- this looks ready to merge as soon as your CLA is signed.

awoods commented on fcrepo-client-impl/src/main/java/org/fcrepo/client/impl/FedoraRepositoryImpl.java:
```diff
@@ -315,8 +327,117 @@ public boolean isWritable() {
+            }
+
+        } catch (final Exception e) {
+            LOGGER.error("could not encode URI parameter", e);
```
Please do not dump exception stacktraces into the log. Instead, print the `e.getMessage()`.

awoods commented on fcrepo-client-impl/src/main/java/org/fcrepo/client/impl/FedoraRepositoryImpl.java:
```diff
@@ -315,8 +327,117 @@ public boolean isWritable() {
+            }
+            LOGGER.debug("Committed transaction {}", TX_ID.get());
+        } catch (final Exception e) {
+            LOGGER.error("could not encode URI parameter", e);
```
Please do not dump exception stacktraces into the log. Instead, print the `e.getMessage()`.

awoods commented on fcrepo-client-impl/src/main/java/org/fcrepo/client/impl/FedoraRepositoryImpl.java:
```diff
@@ -315,8 +327,117 @@ public boolean isWritable() {
+
+            LOGGER.debug("Rolled back transaction {}", TX_ID.get());
+        } catch (final Exception e) {
+            LOGGER.error("could not encode URI parameter", e);
```
Please do not dump exception stacktraces into the log. Instead, print the `e.getMessage()`.

awoods commented on fcrepo-client-impl/src/main/java/org/fcrepo/client/impl/FedoraRepositoryImpl.java:
```diff
@@ -143,7 +153,7 @@ public FedoraDatastream createDatastream(final String path, final FedoraContent
 
             if (statusCode == SC_CREATED) {
-                return getDatastream(path);
+                return getDatastream(prependTransactionId(path));
```
Although this scenario is recognized and ignored in the implementation of `prependTransactionId(..)`, is there ever a case where this apparently redundant call to `prependTransactionId(..)` is not redundant? If not, I would suggest removing the update to this line.

awoods replied:
To clarify, `getDatastream(path)` invokes `prependTransactionId()` on the arg `path`:
https://github.com/fcrepo4-labs/fcrepo4-client/pull/31/files#diff-c56ffc3d56d2f95d81b3208a7d18e4f1R137

Therefore, `getDatastream(prependTransactionId(path))` seems to be redundantly calling `prependTransactionId()`. no?

awoods commented on fcrepo-client-impl/src/main/java/org/fcrepo/client/impl/FedoraRepositoryImpl.java:
```diff
@@ -175,7 +185,7 @@ public FedoraDatastream createOrUpdateRedirectDatastream(final String path, fina
 
             if (statusCode == SC_CREATED) {
-                return getDatastream(path);
+                return getDatastream(prependTransactionId(path));
```
Although this scenario is recognized and ignored in the implementation of `prependTransactionId(..)`, is there ever a case where this apparently redundant call to `prependTransactionId(..)` is not redundant? If not, I would suggest removing the update to this line.

awoods commented on fcrepo-client-impl/src/main/java/org/fcrepo/client/impl/FedoraRepositoryImpl.java:
```diff
@@ -202,7 +212,7 @@ public FedoraObject createObject(final String path) throws FedoraException {
 
             if (statusCode == SC_CREATED) {
-                return getObject(path);
+                return getObject(prependTransactionId(path));
```
See: https://github.com/fcrepo4-labs/fcrepo4-client/pull/31/files#r42413097

awoods commented on fcrepo-client-impl/src/main/java/org/fcrepo/client/impl/FedoraRepositoryImpl.java:
```diff
@@ -252,18 +264,18 @@ public FedoraObject createResource(final String containerPath) throws FedoraExce
     public FedoraDatastream findOrCreateDatastream(final String path) throws FedoraException {
         try {
-            return getDatastream(path);
+            return getDatastream(prependTransactionId(path));
```
See: https://github.com/fcrepo4-labs/fcrepo4-client/pull/31/files#r42413097

awoods commented on fcrepo-client-impl/src/main/java/org/fcrepo/client/impl/FedoraRepositoryImpl.java:
```diff
@@ -252,18 +264,18 @@ public FedoraObject createResource(final String containerPath) throws FedoraExce
+            return getDatastream(prependTransactionId(path));
         } catch ( NotFoundException ex ) {
-            return createDatastream(path, null);
+            return createDatastream(prependTransactionId(path), null);
```
See: https://github.com/fcrepo4-labs/fcrepo4-client/pull/31/files#r42413097

awoods commented on fcrepo-client-impl/src/main/java/org/fcrepo/client/impl/FedoraRepositoryImpl.java:
```diff
@@ -252,18 +264,18 @@ public FedoraObject createResource(final String containerPath) throws FedoraExce
     public FedoraObject findOrCreateObject(final String path) throws FedoraException {
         try {
-            return getObject(path);
+            return getObject(prependTransactionId(path));
```
See: https://github.com/fcrepo4-labs/fcrepo4-client/pull/31/files#r42413097

awoods commented on fcrepo-client-impl/src/main/java/org/fcrepo/client/impl/FedoraRepositoryImpl.java:
```diff
@@ -252,18 +264,18 @@ public FedoraObject createResource(final String containerPath) throws FedoraExce
+            return getObject(prependTransactionId(path));
         } catch ( NotFoundException ex ) {
-            return createObject(path);
+            return createObject(prependTransactionId(path));
```
See: https://github.com/fcrepo4-labs/fcrepo4-client/pull/31/files#r42413097

awoods commented:
Thanks for this, @gushakov. Could you please let me know what your https://jira.duraspace.org user-id is so that we can associate you to this ticket:
https://jira.duraspace.org/browse/FCREPO-1762

awoods commented:
This work was completed with: https://github.com/fcrepo4-labs/fcrepo4-client/pull/36
